### ISSUE TITLE
groupAssignees in roles are ignored for LDAP groups

### ISSUE THREAD

**Issue by res13** [Layer7-Community/graphman-client#66]
I exported all roles from my 11.1.1.18484 source Gateway with the command:
```shell
./graphman.sh export --using roles --output gateway-roles.json
``` 
My source gateway has groupAssignees to two LDAP groups (lets call them group1 and group2) in the role "Administrator". The exported json for the role "Administrator" looks like this:

```javascript
"roles": [
    {
      "goid": "0000000000000000ffffffffffffff9c",
      "name": "Administrator",
      "roleType": "SYSTEM",
      "description": "Users assigned to the {0} role have full access to the gateway.",
      "tag": "ADMIN",
      "replaceAssignees": false,
      "userAssignees": [
        {
          "name": "gmu",
          "login": "gmu",
          "providerName": "Internal Identity Provider",
          "providerType": "INTERNAL"
        },
        {
          "name": "ssm_admin",
          "login": "ssm_admin",
          "providerName": "Internal Identity Provider",
          "providerType": "INTERNAL"
        }
      ],
      "groupAssignees": [
        {
          "name": "group1",
          "subjectDn": "CN=group1,OU=Roles,OU=Groups,OU=RBAC,DC=my,DC=company,DC=com",
          "providerName": "ldap.my.company.com",
          "providerType": "LDAP"
        },
        {
          "providerName": "ldap.my.company.com",
          "providerType": "LDAP"
        },
        {
          "name": "group2",
          "subjectDn": "CN=group2,OU=Roles,OU=Groups,OU=RBAC,DC=my,DC=company,DC=com",
          "providerName": "ldap.my.company.com",
          "providerType": "LDAP"
        }
      ]
    },
    [...]
]    
``` 

When I try to import the roles to my target gateway (also 11.1.1.18484) with:
```shell
./graphman.sh import --using roles --input gateway-roles.json
``` 

it first complained that there is a groupAssignement without a name/subjectDn. I noticed that it is this one:
```javascript
        {
          "providerName": "ldap.my.company.com",
          "providerType": "LDAP"
        }
``` 

I don't know why this part is there (because there are only two groupAssignements in my source gateway), so I removed it manually from the file and tried again. This time the import finishes without any errors but it also totally ignores the groupAssignees. The console output of the "Administrator" role looks like this:
```shell
{
        "goid": "0000000000000000ffffffffffffff9c",
        "name": "Administrator",
        "checksum": "7e5141898df7c219f1e0d105e473f3faae4f723c",
        "roleType": "SYSTEM",
        "description": "Users assigned to the {0} role have full access to the gateway.",
        "tag": "ADMIN",
        "replaceAssignees": false,
        "userAssignees": [
          {
            "name": "gmu",
            "login": "gmu",
            "providerName": "Internal Identity Provider",
            "providerType": "INTERNAL"
          },
          {
            "name": "ssm_admin",
            "login": "ssm_admin",
            "providerName": "Internal Identity Provider",
            "providerType": "INTERNAL"
          }
        ],
        "groupAssignees": []
      }
``` 
Also when I check the target gateway with the policy manager, the groupAssignees are not there.

In conclusion I have two issues:
1. The export of roles returns an unnecessary and incomplete groupAssignement
2. The import of roles ignores the groupAssignees to LDAP groups in general

**Comment by graju256** [comment]
@res13 BTW, I'm curious about filtering unwanted roles (where no user and group assignees). We've covered this use-case but went into the product with no documentation. 

Use the below option to discard the unwanted roles being captured during the **export** operation
` --options.excludeRolesIfRequired`

Thanks.

**Comment by res13** [comment]
> @res13
> 
> Q1) Not sure, a group assignment without **name**? Could you please inspect the **Administrator** role from the policy manager?

The Administrator role of the source gateway lookes ok to me in the policy manager (two internal users and two ldap groups):
![image](https://github.com/user-attachments/assets/7cd90f72-af1f-468e-8ff1-dc1b09e8a0ad)

But I also checked the output of the "old" restman https://source-gateway-host:8443/restman/1.0/roles/0000000000000000ffffffffffffff9c and there I see a third assignement to a LDAP group which does not exist anymore (was deleted on LDAP). I used restman to delete this assignement and now graphman does not show it either. So problem 1 solved.

> Q2) Import operation suppose to fail if any one of the group assignees fails to resolve. As you've specified the --using roles option, mutation cannot be done. You should be using --using install-bundle. Actually, install-bundle is the default mutation query. Hence, this option can be ignored.

When I use the default mutation query I get the following output for each role: 
```javascript
          "action": "NEW_OR_UPDATE",
          "status": "ERROR",
          "description": "Error creating or updating RBAC_ROLE entity, Creating a role is unsupported"
``` 

> client should have checked the query and warn the user if it is not mutation type. We will fix this in future releases by issuing the appropriate error. Fortunately, we've taken care of accidental mutations (i.e., export command paired with mutation based queries).

nice, thanks 😄 

> @res13 BTW, I'm curious about filtering unwanted roles (where no user and group assignees). We've covered this use-case but went into the product with no documentation.
> 
> Use the below option to discard the unwanted roles being captured during the **export** operation ` --options.excludeRolesIfRequired`
> 
> Thanks.

Neat feature. With this option, only the "Administrator" role gets exported and it also solves my problem 2 somehow. I don't really understand why, but as long as it works I am happy 😄 

conclusion: both problems solved, thanks 💯

**Comment by graju256** [comment]
@res13 New role creation is not supported, hence it leads to mutation failure. Fortunately, options.excludeRolesIfRequired option filtered them out, hence there were no errors.

**Conclusion**: migrating/creating the new roles are not supported.
